CHANGELOG.md

```diff
@@ -4,8 +4,9 @@ dev
  - Fix bug when attempting to parse function with a magic scope parameter type hint (eg `function (self $param) { ... }`.
  - Implement `Analysis\TolerantExpressionAnalyser` which will convert analysis exceptions into the *mixed* type.
  - Add `Providers\DSL\Compilation\Parameters\ParameterCollection::contains` to check whether the collection contains a parameter.
- - Add `Providers\DSL\Compilation\Parameters\ParameterCollection::remove` to remove a previously added parameter.\
+ - Add `Providers\DSL\Compilation\Parameters\ParameterCollection::remove` to remove a previously added parameter.
  - Introduce `Analysis\INativeType::TYPE_NUMERIC` acting as a union type for `Analysis\INativeType::TYPE_INT` and `Analysis\INativeType::TYPE_DOUBLE`  
+ - Change identity hash to hash instances of `DateTimeInterface` by class and timestamp hence treating it as a value type throughout the PINQ API
```

Comment:
@Ocramius Hmm.. this is a tough one, perhaps some sort of static registrar defining hasher implementations for a hasher interface for a specific class? The main reason I am considering `DateTime` as a special case is for compatibility with SQL where date times are a value type. For instance:

```php
$results = $table
        ->groupBy(function ($row) { return $row['some_date']; })
        ->select(function ($group, \DateTime $date) {
               return [
                    'date'  => $date,
                    'count' => $group->count()
              ];
        });
```
Where `$row['some_date']` is a datetime in the database represented as an instance of `DateTime` in the query, this will not map well to SQL as it stands as in fact each row containing an instance of `DateTime` will produce a new group. This update fixes that as the behaviour of the API now matches SQL.

As you have mentioned, perhaps it is worthwhile implementing way for users to customize this behaviour on a per-class approach. The trouble I have with this idea is again the compatibility with SQL, when mapping entities to rows this is all well and good because what is the instance in PHP corresponds to the primary key in SQL. For value objects, this is more difficult as there is no language level concept in PHP, if we define a way for users to register a special hasher for their classes, it will have to be in some formalized manner that can be interpreted by a query provider. I don't believe it would be a good idea just to provide an interface as such:

```php
interface IClassIdentifier
{
    public function getClassType(): string;
    public function getIdentityString($instance): string;
}
```

There would be no way for any query provider to interpret such an implementation without inspecting the implementation of `getIdentityString` method which is just nasty. The reason I bring all this this up is because I have begun implementing a query provide like Pinq.Demo.Sql but a much more thorough implementation using the Doctrine so this is at the forefront of my mind. Will have to think this over some more, thoughts?